x/slashing/keeper/keeper_test.go

```diff
@@ -2,346 +2,79 @@ package keeper_test
-
-	addr, val := pks[0].Address(), pks[0]
-	consAddr := sdk.ConsAddress(addr)
-	tstaking := teststaking.NewHelper(s.T(), ctx, s.stakingKeeper)
-	valAddr := sdk.ValAddress(addr)
-
-	tstaking.CreateValidatorWithValPower(valAddr, val, power, true)
-	validatorUpdates := staking.EndBlocker(ctx, s.stakingKeeper)
-	s.Require().Equal(2, len(validatorUpdates))
-	tstaking.CheckValidator(valAddr, stakingtypes.Bonded, false)
-
-	// 100 first blocks OK
-	height := int64(0)
-	for ; height < int64(100); height++ {
-		ctx = ctx.WithBlockHeight(height)
-		s.slashingKeeper.HandleValidatorSignature(ctx, val.Address(), power, true)
-	}
-
-	// kick first validator out of validator set
-	tstaking.CreateValidatorWithValPower(sdk.ValAddress(pks[1].Address()), pks[1], power+1, true)
-	validatorUpdates = staking.EndBlocker(ctx, s.stakingKeeper)
-	s.Require().Equal(2, len(validatorUpdates))
-	tstaking.CheckValidator(sdk.ValAddress(pks[1].Address()), stakingtypes.Bonded, false)
-	tstaking.CheckValidator(valAddr, stakingtypes.Unbonding, false)
-
-	// 600 more blocks happened
-	height = height + 600
-	ctx = ctx.WithBlockHeight(height)
-
-	// validator added back in
-	tstaking.DelegateWithPower(sdk.AccAddress(pks[2].Address()), valAddr, 50)
-
-	validatorUpdates = staking.EndBlocker(ctx, s.stakingKeeper)
-	s.Require().Equal(2, len(validatorUpdates))
-	tstaking.CheckValidator(valAddr, stakingtypes.Bonded, false)
-	newPower := power + 50
-
-	// validator misses a block
-	s.slashingKeeper.HandleValidatorSignature(ctx, val.Address(), newPower, false)
-	height++
-
-	// shouldn't be jailed/kicked yet
-	tstaking.CheckValidator(valAddr, stakingtypes.Bonded, false)
-
-	// validator misses an additional 500 more blocks, after the cooling off period of SignedBlockWindow (here 1000 blocks).
-	latest := s.slashingKeeper.SignedBlocksWindow(ctx) + height
-	for ; height < latest+s.slashingKeeper.MinSignedPerWindow(ctx); height++ {
-		ctx = ctx.WithBlockHeight(height)
-		s.slashingKeeper.HandleValidatorSignature(ctx, val.Address(), newPower, false)
-	}
-
-	// should now be jailed & kicked
-	staking.EndBlocker(ctx, s.stakingKeeper)
-	tstaking.CheckValidator(valAddr, stakingtypes.Unbonding, true)
-
-	// check all the signing information
-	signInfo, found := s.slashingKeeper.GetValidatorSigningInfo(ctx, consAddr)
-	s.Require().True(found)
-	s.Require().Equal(int64(700), signInfo.StartHeight)
-	s.Require().Equal(int64(499), signInfo.MissedBlocksCounter)
-	s.Require().Equal(int64(499), signInfo.IndexOffset)
-
-	// some blocks pass
-	height = int64(5000)
-	ctx = ctx.WithBlockHeight(height)
-
-	// validator rejoins and starts signing again
-	s.stakingKeeper.Unjail(ctx, consAddr)
-
-	s.slashingKeeper.HandleValidatorSignature(ctx, val.Address(), newPower, true)
-
-	// validator should not be kicked since we reset counter/array when it was jailed
-	staking.EndBlocker(ctx, s.stakingKeeper)
```

Comment:
Hey quick question @cool-develope : did you also convert this test to use mocks? In particular how did you deal with this EndBlocker function, that takes a real keeper and not an interface.

(having the same issue in x/gov)

Reply:
no, I couldn't mock that one, this is why I moved `keeper_test.go` to integration/slashing

Reply:
@amaurym tests/integration/slashing/keeper/keeper_test.go